tests/Client.hs

```diff
@@ -0,0 +1,232 @@
+{-# LANGUAGE BinaryLiterals #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-- | This modules provdes a way to emulate certain client behaviour for testing
+-- purposes. It DOES NOT implement the webauthn specification for two reasons:
+-- 1. There is no need for it in our tests.
+-- 2. It is much more convenient to implement both the client and authenticator
+-- in a single function, removing their communication.
+module Client (clientAssertion, spec) where
+
+import Authenticator
+  ( Authenticator (AuthenticatorNone, aCredentials, aSupportedAlgorithms),
+    authenticatorGetAssertion,
+    authenticatorMakeCredential,
+  )
+import qualified Client.PrivateKey as PrivateKey
+import qualified Crypto.Fido2.Model as M
+import qualified Crypto.Fido2.Model.JavaScript as JS
+import Crypto.Fido2.Model.JavaScript.Decoding
+  ( decodeCreatedPublicKeyCredential,
+    decodePublicKeyCredentialCreationOptions,
+    decodePublicKeyCredentialRequestOptions,
+    decodeRequestedPublicKeyCredential,
+  )
+import Crypto.Fido2.Model.JavaScript.Encoding
+  ( encodeCreatedPublicKeyCredential,
+    encodePublicKeyCredentialCreationOptions,
+    encodePublicKeyCredentialRequestOptions,
+    encodeRequestedPublicKeyCredential,
+  )
+import qualified Crypto.Fido2.Model.JavaScript.Types as JS
+import Crypto.Fido2.Operations.Attestation (allSupportedFormats)
+import qualified Crypto.Fido2.Operations.Attestation as Fido2
+import qualified Crypto.Fido2.PublicKey as PublicKey
+import Crypto.Hash (hash, hashlazy)
+import Crypto.Random (MonadRandom)
+import Data.Aeson (encode)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Base64.URL as Base64
+import Data.Either (fromRight, isRight)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe)
+import qualified Data.Set as Set
+import Data.Text.Encoding (decodeUtf8)
+import Data.Validation (toEither)
+import System.Random.Stateful (globalStdGen, uniformM)
+import Test.Hspec (SpecWith, describe, it, shouldSatisfy)
+
+-- | Emulates the client-side operation for attestation given an authenticator.
+-- MonadRandom is required during the geneation of the new credentials.
+clientAttestation :: (MonadRandom m, MonadFail m) => JS.PublicKeyCredentialCreationOptions -> Authenticator -> m (JS.CreatedPublicKeyCredential, Authenticator)
+clientAttestation options authenticator = do
+  let M.PublicKeyCredentialCreationOptions {..} =
+        fromRight (error "Test: could not decode creation options") $ decodePublicKeyCredentialCreationOptions options
+      clientDataBS =
+        encode
+          JS.ClientDataJSON
+            { JS.typ = "webauthn.create",
+              JS.challenge = decodeUtf8 . Base64.encode $ M.unChallenge pkcocChallenge,
+              JS.origin = "https://localhost:8080",
```

Comment:
As with the authenticator, there should be a parameter for the origin/domain/relying party that we're connected to